### ISSUE TITLE
Spike: Start looking at buffered communication microbenchmarks

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#9652]
There are some buffered comm / histogram benchmarks where we're lagging behind reference UPC and maybe shmem versions. Start digging into these to get a better understanding of where/why our performance lags.

Reference versions are at https://github.com/jdevinney/bale. Here's a UPC example:

```c
#pragma pgas defer_sync
  for(i = 0; i < T; i++) {
    lgp_atomic_add(counts, index[i], -v);
  }
  lgp_barrier();
```

where I imagine the main reason we're behind is because the `#pragma pgas defer_sync` results in non-blocking operations and our atomic operations are always blocking today. We'll have to do some performance studies to find out for sure though.

**Comment by ronawho** [comment]
Here's what I'm using to build the reference version so far:

```sh
git clone git@github.com:jdevinney/bale.git
cd bale

# setup cray compiler as the upc compiler
module unload $(module list -t 2>&1 | grep PrgEnv-)
module load PrgEnv-cray
export UPC=cc
export UPCFLAGS="-hupc"

# configure scripts seem to be missing execute permission
find . -name configure | xargs chmod +x 

# -f forces autoconf remake (configure scripts don't work out of the box for me)
# -c just passes flags onto the configure scripts
export PLATFORM=cray
sh install.sh -f -c CC=cc
```

And to run the histogram code on 28-core (56 thread) broadwell compute nodes: 

```sh
export PLATFORM=cray
export CORES_PER_NODE=28
export NODES=2

srun --nodes=$NODES --ntasks=$(($NODES*CORES_PER_NODE)) --ntasks-per-node=$CORES_PER_NODE --constraint=BW28 ./build_$PLATFORM/bin/histo -M 1
```

Results for 2, 4, 8, 16, and 32 nodes all seem to be around ~4.2 seconds.

**Comment by ronawho** [comment]
Here's a maybe equivalent Chapel version of [histo_agi](https://github.com/jdevinney/bale/blob/master/apps/histo_src/histo_agi.upc): 

```chpl
use Time;
use CyclicDist;
use BlockDist;
use Random;

config const N=2000000 * here.maxTaskPar; // number of updates
config const M=1000 * here.maxTaskPar * numLocales; // size of table

 // allocate main table and array of random ints
const Mspace = {0..M-1};
const D = Mspace dmapped Cyclic(startIdx=Mspace.low);
var A: [D] atomic int;

const Nspace = {0..(N*numLocales - 1)};
const D2 = Nspace dmapped Block(Nspace);
var rindex: [D2] int;

/* set up loop */
fillRandom(rindex, 208); // the 208 is a seed
forall r in rindex {
 r = mod(r, M);
}

var t: Timer;
t.start();
/* main loop */
forall r in rindex {
  A[r].add(1); //atomic add
}
t.stop();
writeln(t.elapsed());
```

Compiled with `--fast` and using gcc 6.3 as the backend compiler I see timings between 4.8 and 5.6 seconds for 2, 4, 8, 16, and 32 locales/nodes.

I had expected performance to be significantly off, but that doesn't seem to be the case. Maybe I'm not running something quite right.

**Comment by ronawho** [comment]
Looks like the reference code didn't quite result in NB atomics. It should be something like: 

```c
 for(i = 0; i < T; i++) {
#pragma pgas defer_sync
    counts[index[i]] += -v;
  }
  lgp_barrier();
```

With that version I see around 1.2s (down from 4.2s) with the Chapel version still at ~5.4s. So it seems like the key reason we're losing is that we don't currently have support for non-blocking atomic operations. I think we have the language level support for this (just use relaxed memory ordering), but we don't have runtime support for it. I think it'd look something like

```chpl 
forall r in rindex {
  A[r].add(1, order=memory_order_relaxed); //atomic add
}
```

I'll start digging into the ugni support of figure out how hard that will be to implement. I think issuing non-blocking atomics will be easy, the hard part will be the code to ensure we wait for them to complete at a fence (or the implicit fence/sync at the end of a forall/coforall)

**Comment by ronawho** [comment]
I believe this initial investigation is done, and we know why we're behind (lack of non-blocking atomics.) 

We're 25% behind the blocking reference version and 4-5x off of the non-blocking version. I think the next steps are to start to explore how much work will be involved for adding non-blocking atomics and possibly looking at speeding up our blocking atomics (i.e. why are we 25% behind the blocking version)

I've opened a new epic to track this: https://github.com/chapel-lang/chapel/issues/9782 and I've added most of the information from this thread to that.
